Prevent private key extraction from block signatures

hewison-chris opened the pull request:
The noise for the signature was being created by adding the preimage to
enrollment commitment secret. This was flawed as with two signatures the
private key could be extracted. The solution is to multiple the scaler
preimage by the secret enrollment commitment to generate the signature
noise.

hewison-chris committed:
Prevent private key extraction from block signatures

The noise for the signature was being created by adding the preimage to
enrollment commitment secret. This was flawed as with two signatures the
private key could be extracted. The solution is to multiply the scaler
preimage by the secret enrollment commitment to generate the signature
noise.

hewison-chris commented:
![updated block signature noise 001](https://user-images.githubusercontent.com/11815575/126922457-2cbaa4c9-8439-40a6-aaa3-b66c4c159b71.jpeg)

Updated signature noise calculation for signing blocks and verifying them.

linked0 commented on source/agora/node/Ledger.d:
```diff
@@ -888,7 +888,7 @@ public class Ledger
                 return "Block: Couldn't find pre-image for validator";
             }
-            Point R = CR + Scalar(preimage_info.hash).toPoint();
+            Point R = Scalar(preimage_info.hash) * CR;
```
```suggestion
            Point R = CR * Scalar(preimage_info.hash);
```
Better to change the order of the operands as other statements.

hewison-chris replied:
I applied your change as it then matches the slide.

linked0 reviewed (approved):
LGTM. One suggestion which is not so really important. :)

hewison-chris commented:
The `UnlockAge` unit test was causing failures for Ubuntu ci runs. I added the first commit to address making the test more robust.

hewison-chris commented:
Now the `module agora.test.ValidatorRecurringEnrollment` unit test is also showing a race condition. Added second commit to prevent this one.

Geod24 commented:
Can you submit the first commit as its own PR ?

Geod24 commented:
Or both fixes

kchulj reviewed (approved):
LGTM
Typo in 3rd line of 3rd commit message

hewison-chris commented:
> LGTM
> Typo in 3rd line of 3rd commit message

Good catch, I'll update.

hewison-chris committed:
wip - potential issue still

The unit test in data.Block demonstrates that it is still not secure so
it looks like we will have to use a random r for signature noise.

hewison-chris commented:
The last commit shows we still need to update the noise used for signing.

hewison-chris commented:
Closing as no longer relevant after recent block signing signature scheme updates.
